api/samples/inscount.cpp

```diff
@@ -164,9 +164,36 @@ event_bb_analysis(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
             }
         }
     }
-    /* Count instructions */
-    for (instr = instrlist_first_app(bb), num_instrs = 0; instr != NULL;
-         instr = instr_get_next_app(instr)) {
+
+    /* Count instructions. If an emulation client is running with this client,
+     * we want to count all the original native instructions and the emulated
+     * instruction but NOT the introduced native instructions used for emulation.
+     */
+    bool is_emulation = false;
+    for (instr = instrlist_first(bb), num_instrs = 0; instr != NULL;
+         instr = instr_get_next(instr)) {
```

Comment:
An alternative for having the is_emulation bool might be:

```
    for (instr = instrlist_first(bb), num_instrs = 0; instr != NULL;
         instr = instr_get_next(instr)) {
         if (drmgr_is_emulation_start(instr)) {
               /* Skip emulation instructions */
             while (!drmgr_is_emulation_end(instr)) {
                  ASSERT(instr_get_next(instr) != NULL, "no emulation_end found");
                  instr = instr_get_next(instr);
             }
         } else if (!instr_is_app(instr)) {
            /* Skip other non-app instructions. Note that we count a block of emulation (non-app) instructions as a simple app instruction. */
            continue;
        }
         num_instrs++; 
    }
```

Personally I think that is slightly easier to reason about, but

Reply:
Personally, I don't find this any easier to reason about :-) but it is more compact.

Reply:
When this patch is merged, I'm planning to do a sample use case and may change the loop then.